### PR TITLE
Feature/round results

### PR DESCRIPTION
**Pull request by rjs68**
-Added toString method in RoundOutcome that displays the result of the round
-Changed the onRoundEnd method to take advantage of the new toString method
-Created additional onRoundEnd methods to account for and display and players that have been eliminated from the game

### CODE REVIEW

**Comment by hughwin** [comment]
Added a commit on top to satisfy the story: "As a human player, when it is not my turn I want to be able to select next so that I can see the results of each round (2pts)".

File: src/main/java/com/toptrumps/core/engine/Game.java [hunk 1]
```diff
@@ -43,7 +43,7 @@ public Game(int numberOfPlayers, GameEventListener listener) {
             add(new Player(0, DEFAULT_USER_NAME));
         }};
 
-        createAIPlayers(numberOfPlayers - DEFAULT_NUMBER_OF_HUMAN_PLAYERS);
+        createAIPlayers(numberOfPlayers);
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/core/engine/Game.java]: Oooops, sorry about that!

File: src/main/java/com/toptrumps/core/engine/Game.java [hunk 2]
```diff
@@ -100,8 +100,11 @@ public void start() {
             // TODO: Revisit this to avoid another loop
             players.forEach(Player::removeTopCard);
 
+            Player previousActivePlayer = null;
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/core/engine/Game.java]: Do we really need this?

File: src/main/java/com/toptrumps/core/engine/GameEventListener.java [hunk 3]
```diff
@@ -12,6 +12,8 @@
     Attribute onRequestSelection(final Card card);
     void onAttributeSelected(final Player activePlayer);
     void onRoundEnd(final RoundOutcome outcome);
+    void onRoundEnd(final RoundOutcome outcome, final Player eliminatedPlayer);
+    void onRoundEnd(final RoundOutcome outcome, final ArrayList<Player> eliminatedPlayers);
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/core/engine/GameEventListener.java]: These are supposed to be life cycle methods. It's very confusing to have overloads in this case! We should keep online one "onRoundEnd" and pass to it everything we might need. If we have +3 parameters, then it's time to create an object.

File: src/main/java/com/toptrumps/core/engine/GameEventListener.java [hunk 4]
```diff
@@ -12,6 +12,8 @@
...
     void onGameOver(Player winner);
-
+    void onAiRoundWin();
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/core/engine/GameEventListener.java]: We certainly don't need this! ```onRoundEnd``` can return the correct winner (doesn't matter if it's AI or not).

File: src/main/java/com/toptrumps/core/engine/Game.java [hunk 5]
```diff
@@ -116,7 +119,36 @@ public void start() {
...
+            //Call the correct method for the round end 
+            if(eliminatedPlayer == null && eliminatedPlayers.isEmpty()){
+                listener.onRoundEnd(outcome);
+            }else if(eliminatedPlayers.isEmpty()){
+                listener.onRoundEnd(outcome, eliminatedPlayer);
+            }else{
+                listener.onRoundEnd(outcome, eliminatedPlayers);
+            }
+
+            if(previousActivePlayer instanceof AIPlayer){
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/core/engine/Game.java]: We have a convenience method on ```Player``` to determine if it's AI or not.

File: src/main/java/com/toptrumps/core/engine/Game.java [hunk 6]
```diff
@@ -116,7 +119,36 @@ public void start() {
...
             }
-            listener.onRoundEnd(outcome);
+
+            //Determine if any players have been eliminated
+            Player eliminatedPlayer = null;
+            ArrayList<Player> eliminatedPlayers = new ArrayList<Player>();
+            for(Player player: players){
+                if(player.getDeck().isEmpty()){
+                    if(eliminatedPlayer == null){
+                        eliminatedPlayer = player;
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/core/engine/Game.java]: We can get the eliminated player from just filtering players with empty decks. It's a minor tweak in the current logic, like this:

```
   ArrayList<Player> defeatedPlayers = (ArrayList<Player>) players.stream().filter(p -> p.getDeck().isEmpty()).collect(toList());
            players.removeAll(defeatedPlayers);
```

File: src/main/java/com/toptrumps/core/engine/RoundOutcome.java [hunk 7]
```diff
@@ -40,4 +40,26 @@ public Player getWinner() {
         return draws;
     }
 
+    public String toString(){
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/core/engine/RoundOutcome.java]: This is presentation logic. Should be done in the Controller.

**Review by BelfDev — CHANGES_REQUESTED** [review]
# Summary

- Please eliminate different ```onRoundEnd``` signatures
- Please revise the ```eliminatedPlayers``` logic
- Please double-check if we already have methods implemented (no need to use ```instanceof AIPlayer```
- Please remove ```onAiRoundWin()``` and use ```onRoundEnd```
- Please keep the presentation logic in the Controller/View

File: src/main/java/com/toptrumps/cli/CommandLineController.java [hunk 8]
```diff
@@ -80,17 +81,62 @@ public void onAttributeSelected(Player activePlayer) {
...
+        showRoundResult(outcome);
+        ArrayList<Player> removedPlayers = outcome.getRemovedPlayers();
+        if(!removedPlayers.isEmpty()){
+            showEliminatedPlayers(removedPlayers);
+        }
+    }
+
+    private void showRoundResult(RoundOutcome outcome){
+        RoundOutcome.Result result = outcome.getResult();
+        String roundResult = "\nThe round resulted in a: ";
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/cli/CommandLineController.java]: Above you have a ```RoundOutcome.Result``` object which you named ```result```. Just beneath you have a ```String``` named ```roundResult```. You see the issue here? Semantically, what's truly the result?

I think ```String roundResult``` is in fact your ```outcomeMessage```.

File: src/main/java/com/toptrumps/cli/CommandLineController.java [hunk 9]
```diff
@@ -80,17 +81,62 @@ public void onAttributeSelected(Player activePlayer) {
...
+        switch(result){
+            case VICTORY:
+                roundResult += "Victory!\nThe winner is: " + outcome.getWinner().getName();
+                break;
+            case DRAW:
+                roundResult += "Draw\nThe score was tied between: ";
+                for(Player player: outcome.getDraws()){
+                    if(player == outcome.getDraws().get(outcome.getDraws().size()-1)){
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/cli/CommandLineController.java:101]: Look at this ```if``` statement. How can we make it clearer?

File: src/main/java/com/toptrumps/cli/CommandLineController.java [hunk 10]
```diff
@@ -80,17 +81,62 @@ public void onAttributeSelected(Player activePlayer) {
...
+                    }else{
+                        roundResult += player.getName() + ", ";
+                    }
+                }
+                break;
+        }
+        System.out.println(roundResult);
+    }
+
+    private void showEliminatedPlayers(ArrayList<Player> removedPlayers){
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/cli/CommandLineController.java]: ```showEliminatedPlayers``` received "removedPlayers" => Again, watch the semantics.

File: src/main/java/com/toptrumps/core/engine/Game.java [hunk 11]
```diff
@@ -100,28 +100,29 @@ public void start() {
             // TODO: Revisit this to avoid another loop
             players.forEach(Player::removeTopCard);
 
+            //collect any defeated players and remove them from the game
+            ArrayList<Player> defeatedPlayers = (ArrayList<Player>) players.stream().filter(p -> p.getDeck().isEmpty()).collect(toList());
+            players.removeAll(defeatedPlayers);
+            numberOfPlayers = players.size();
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/core/engine/Game.java:106]: Sweeeeet!

File: src/main/java/com/toptrumps/core/engine/RoundOutcome.java [hunk 12]
```diff
@@ -6,26 +6,28 @@
 
 public class RoundOutcome {
 
-    enum Result {
+    public enum Result {
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/core/engine/RoundOutcome.java:9]: I might refactor this later

File: src/main/java/com/toptrumps/core/engine/RoundOutcome.java [hunk 13]
```diff
@@ -6,26 +6,28 @@
...
         VICTORY, DRAW
     }
 
     private Result result;
     private Player winner;
     private ArrayList<Player> draws;
+    private ArrayList<Player> removedPlayers;
 
-    private RoundOutcome(Result result, Player winner, ArrayList<Player> draws) {
+    private RoundOutcome(Result result, Player winner, ArrayList<Player> draws, ArrayList<Player> removedPlayers) {
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/core/engine/RoundOutcome.java:18]: GREAT catch!!!! 10/10

File: src/main/java/com/toptrumps/core/engine/RoundOutcome.java [hunk 14]
```diff
@@ -40,4 +42,8 @@ public Player getWinner() {
         return draws;
     }
 
+    public ArrayList<Player> getRemovedPlayers(){
```
Comment (BelfDev) [inline src/main/java/com/toptrumps/core/engine/RoundOutcome.java:45]: Removed or defeated? Let's choose something and stick to it.

**Review by BelfDev — CHANGES_REQUESTED** [review]
# Summary

Much better now!!!! Some great catches as well. 

- Please watch the semantics! 
- Please try to make what is going on inside the ```ìf```statements clearer.

**Review by BelfDev — APPROVED** [review]

**Review by nathanieldschneider — APPROVED** [review]

**Review by Arden488 — APPROVED** [review]

**Pull request merged**